src/relay/backend/contrib/dnnl/query_layout.cc

```diff
@@ -0,0 +1,183 @@
+#include <tvm/relay/attrs/nn.h>
+#include <tvm/relay/expr_functor.h>
+#include <tvm/relay/transform.h>
+#include <tvm/relay/type.h>
+#include <tvm/runtime/module.h>
+#include <tvm/runtime/registry.h>
+
+#include <fstream>
+#include <numeric>
+#include <sstream>
+#include "dnnl.hpp"
+
+#include "../../utils.h"
+
+using dim_t = dnnl_dim_t;
+using dims_t = dnnl_dims_t;
+
+namespace tvm {
+namespace relay {
+namespace contrib {
+
+template <typename T, typename U>
+inline void array_set(T *arr, const U &val, size_t size) {
+    for (size_t i = 0; i < size; ++i)
+        arr[i] = static_cast<T>(val);
+}
+
+template <typename T>
+inline void array_copy(T *dst, const T *src, size_t size) {
+    for (size_t i = 0; i < size; ++i)
+        dst[i] = src[i];
+}
+
+void compute_blocks(dims_t &blocks, const dnnl::memory::desc *md) {
+        using format_kind_t = dnnl_format_kind_t;
+        const format_kind_t blocked = dnnl_blocked;
+        if (!(md->data.format_kind==blocked)) {
+            array_set(blocks, 0, md->data.ndims);
+            return;
+        }
+
+        array_set(blocks, 1, md->data.ndims);
+
+        const auto &bd = md->data.format_desc.blocking;
+        for (int iblk = 0; iblk < bd.inner_nblks; ++iblk)
+            blocks[bd.inner_idxs[iblk]] *= bd.inner_blks[iblk];
+    }
+
+bool has_runtime_strides(const dnnl::memory::desc *md) {
+        using format_kind_t = dnnl_format_kind_t;
+        const format_kind_t blocked = dnnl_blocked;
+        if (!(md->data.format_kind==blocked)) return false;
+        for (int d = 0; d < md->data.ndims; ++d)
+            if (md->data.format_desc.blocking.strides[d] == DNNL_RUNTIME_DIM_VAL) return true;
+        return false;
+    }
+
+template <typename T>
+void swap(T &t1, T &t2) {

```

Comment:
mark the function and the above one as `inline`?

Reply:
fixed.